lua/nvim-tree/actions/tree/modifiers/expand-all.lua

```diff
@@ -18,42 +19,56 @@ local function to_lookup_table(list)
 end
 
 ---@param node Node
-local function expand(node)
-  node = lib.get_last_group_node(node)
-  node.open = true
+local function populate_node(node)
+  -- noop if it is a file
+  if node.nodes == nil then
+    return
+  end
   if #node.nodes == 0 then
-    core.get_explorer():expand(node)
+    local cwd = node.link_to or node.absolute_path
+    local handle = vim.loop.fs_scandir(cwd)
+    if not handle then
+      return
+    end
+    local status = git.load_project_status(cwd)
+    core.get_explorer():expand(node, status)
   end
 end
 
 ---@param expansion_count integer
 ---@param node Node
+---@param populate_node function
 ---@return boolean
-local function should_expand(expansion_count, node)
+-- luacheck: push ignore populate_node
+local function expand_until_max_or_empty(expansion_count, node, populate_node)
   local should_halt = expansion_count >= M.MAX_FOLDER_DISCOVERY
   local should_exclude = M.EXCLUDE[node.name]
-  return not should_halt and node.nodes and not node.open and not should_exclude
+  local result = not should_halt and node.nodes and not node.open and not should_exclude
+  return result
 end
+-- luacheck: pop
 
-local function gen_iterator()
+local function gen_iterator(should_expand_fn)
   local expansion_count = 0
 
   return function(parent)
-    if parent.parent and parent.nodes and not parent.open then
-      expansion_count = expansion_count + 1
-      expand(parent)
-    end
+    populate_node(parent)
+    parent.open = true
 
-    Iterator.builder(parent.nodes)
+    Iterator.builder({ parent })
       :hidden()
       :applier(function(node)
-        if should_expand(expansion_count, node) then
+        local should_expand = should_expand_fn(expansion_count, node, populate_node)
+        if should_expand then
           expansion_count = expansion_count + 1
-          expand(node)
+          populate_node(node)
```

Comment:
I'm not quite following the logic here, but it looks like we're not actually making many changes e.g.
* `expand` was renamed to `populate_node`
* unnecessary `should_expand` boolean was extracted
* `get_last_group_node` was moved out of `expand`

Please can you refactor this down to the minimum number of changes necessary. Integration with the existing structure will result in simpler code that is more easily reviewable and maintainable.

I appreciate that we've been through many revisions to get to this point, with complexity from earlier commits removed, however we need to revert some structural changes. Perhaps consider how you'd make these changes from scratch, knowing what you have learned over the course of this PR.

Reply:
> Please can you refactor this down to the minimum number of changes necessary. Integration with the existing structure will result in simpler code that is more easily reviewable and maintainable.

I totally agree with such approach. I think I was able to cut down amount of changes to the bare minimum.

There is one thing that I would like to discuss - changes in the `recursor` function. Before, when the logic was only to expand everything recursively, the condition was to check if `node.open==true` (putting `expansion_count aside`). In my case that is invalid as some nodes have been opened but they shouldn't be traversed further. 

In both cases reusing `should_expand` function results in correct behavior. First of all, this doesn't have to be true in general. Maybe sometimes the logic for `should_recurse` will be different than the logic for `should_expand`. 

Second. So, for each node we call `should_expand` twice, once to check if we should expand and second time to check if we should recurse. Between these calls, if should `should_expand` returns `true`, we call `expand` that changes the state of a given node. This means that users *must not* relay on `node.open` in their `should_expand` implementation as this would result in incorrect behavior. 

I don't have enough experience in developing nvim plugins to judge if this is a problem, but wanted to bring this to your attention. Maybe this should at least be mentioned in the documentation?

Reply:
> In both cases reusing should_expand function results in correct behavior. First of all, this doesn't have to be true in general. Maybe sometimes the logic for should_recurse will be different than the logic for should_expand.

I like the consistency, it's more reason-able and will make maintenance easier.

> This means that users _must not_ relay on `node.open` in their `should_expand` implementation as this would result in incorrect behavior.

So long as we document that we are providing a consistent user experience.

Is `node.nodes` correct; can they test that?